File: 4. Recursion/7. Palindrome/solution.py
# Time Complexity: O(log N)
# Space Complexity: O(log N) (due to the recursion stack) + O(1) (for the auxiliary space) = O(log N)
# Auxiliary Space: O(1)
class Solution:
    def __init__(self):
        self._result = 1

    def get_result(self):
        return self._result

    def checkPalindrome(self, S, left, right):
        if left >= right:
            return
        if S[left] != S[right]:
            self._result = 0
            return
        self.checkPalindrome(S, left + 1, right - 1)

    def isPalindrome(self, S):
        self.checkPalindrome(S, 0, len(S) - 1)
        return self.get_result()

if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        S = input()
        ob = Solution()
        answer = ob.isPalindrome(S)
        print(answer)

File: 4. Recursion/7. Palindrome/solution.js
// Time Complexity: O(N)
// Space Complexity: O(N)
// Auxiliary Space: O(N)

// The below code can correctly check if a given string is a palindrome or not. However, both these approaches have higher space complexity, and for very long strings, it may lead to stack overflow errors.

class Solution {
  isPalindrome(S) {
    const stringLength = S.length;
    if (stringLength <= 1) {
      return 1;
    }
    if (S.charAt(0) === S.charAt(stringLength - 1)) {
      return this.isPalindrome(S.slice(1, stringLength - 1));
    } else return 0;
  }
}

// Using tail recursion

class Solution {
  isPalindrome(S) {
    return this.isPalindromeTailRecursive(S, 1);
  }

  isPalindromeTailRecursive(S, accumulator) {
    const stringLength = S.length;
    if (stringLength <= 1) {
      return accumulator;
    }
    if (S.charAt(0) === S.charAt(stringLength - 1)) {
      return this.isPalindromeTailRecursive(
        S.slice(1, stringLength - 1),
        accumulator
      );
    } else {
      return 0;
    }
  }
}
